### PR TITLE
vim-patch:8.2.{3402,3403}

### PR DESCRIPTION
**Pull request by f380cedric**
Needed to be refactored since function moved around.
I did **NOT** test this patch.
Related to CVE-2021-3770.

### CODE REVIEW

**Comment by f380cedric** [comment]
@vigoux

File: src/nvim/option.c [hunk 1]
```diff
@@ -7488,7 +7489,7 @@ bool tabstop_set(char_u *var, long **array)
         if (cp != end) {
           emsg(_(e_positive));
         } else {
-          emsg(_(e_invarg));
+          emsg(_(e_invarg2), cp);
```
Comment (seandewar) [inline src/nvim/option.c:7492]:
```suggestion
          semsg(_(e_invarg2), cp);
```

File: src/nvim/option.c [hunk 2]
```diff
@@ -7501,7 +7502,7 @@ bool tabstop_set(char_u *var, long **array)
       valcount++;
       continue;
     }
-    emsg(_(e_invarg));
+    emsg(_(e_invarg2), var);
```
Comment (seandewar) [inline src/nvim/option.c:7505]:
```suggestion
    semsg(_(e_invarg2), var);
```

**Comment by f380cedric** [comment]
Btw, I have other vim-patches in the pipeline.
Should I open a PR for each one of them?
I also have N/A patches (for instance 8.2.3409),
should I add them in this PR commit message?

**Comment by seandewar** [comment]
v8.2.3409 is not N/A as the test is still applicable, but if you believe a patch is N/A, then you can add it to the commit message like: https://github.com/neovim/neovim/commit/6a930a9dc4ddf190b528c945f19bc0a0bad2cc29 (preferably to the first commit's message).

**Comment by f380cedric** [comment]
> v8.2.3409 is not N/A as the test is still applicable, but if you believe a patch is N/A, then you can add it to the commit message like: [6a930a9](https://github.com/neovim/neovim/commit/6a930a9dc4ddf190b528c945f19bc0a0bad2cc29) (preferably to the first commit's message).

Ok, I will keep the test.

> > Should I open a PR for each one of them?
> 
> If they're relatively trivial or related you could use the same PR, but maybe it'll be useful so that all patches aren't blocked if there are review comments for the others?

I do not think they are related, there were cherry-picked from vim CVEs.
But I think most of them will be of the same complexity than this one.

Unless advised otherwise, I am going to open a PR per patch.

**Comment by seandewar** [comment]
By the way, I think https://github.com/neovim/neovim/pull/15972/commits/cb3ca707edd82849172738ccd1c5891d63b73f13 is related to a CVE, but it's held up waiting for review with the rest of the PR.
It does require some test changes, so feel free to cherry-pick that one if you need to.

**Comment by f380cedric** [comment]
> By the way, I think [cb3ca70](https://github.com/neovim/neovim/commit/cb3ca707edd82849172738ccd1c5891d63b73f13) is related to a CVE, but it's held up waiting for review with the rest of the PR. It does require some test changes, so feel free to cherry-pick that one if you need to.

Yes, there are 8.2.{3487,3581,3625} for which there are already opened PRs.
Should I cherry-pick them from the PRs? What will happen to the commit of those PRs?
Won't it force the team to do twice the work when reviewing,
i.e reviewing your commit in your multi-commit PR and reviewing your commit in my one-commit PR?

**Comment by seandewar** [comment]
> Won't it force the team to do twice the work when reviewing,
> i.e reviewing your commit in your multi-commit PR and reviewing your commit in my one-commit PR?

I can just remove the commit from my PR. It isn't really related to the other commits there, so there's no point having it be held up waiting for a review with the others.
I can't speak for the other patches though.

**Comment by zeertzjq** [comment]
Moving patches 8.2.3581 and 8.2.3625 to separate PRs may cause merge conflicts in test files.

**Review by seandewar — APPROVED** [review]
Tested and LGTM

File: src/nvim/option.c [hunk 3]
```diff
@@ -7510,7 +7511,16 @@ bool tabstop_set(char_u *var, long **array)
...
   t = 1;
   for (cp = var; *cp != NUL;) {
-    (*array)[t++] = atoi((char *)cp);
+    int n = atoi((char *)cp);
+
+    // Catch negative values, overflow and ridiculous big values.
+    if (n < 0 || n > 9999) {
+      semsg(_(e_invarg2), cp);
+      xfree(*array);
+      *array = NULL;
```
Comment (seandewar) [inline src/nvim/option.c:7520]: bit of a nit actually but you could use `XFREE_CLEAR` here instead

```suggestion
      XFREE_CLEAR(*array);
```

Comment (f380cedric) [reply src/nvim/option.c:7520]: Yeah, I did not understand why upstream was not using `VIM_CLEAR`...

Comment (seandewar) [reply src/nvim/option.c:7520]: feel free to add it and i'll merge :+1:

**Pull request merged**